File: java-tomcat/src/main/java/com/ederfmatos/javatomcat/usecase/CreateTransaction.java
package com.ederfmatos.javatomcat.usecase;

import com.ederfmatos.javatomcat.entity.Transaction;
import com.ederfmatos.javatomcat.enums.Currency;
import com.ederfmatos.javatomcat.gateway.CurrencyGateway;
import com.ederfmatos.javatomcat.repository.TransactionRepository;
import java.math.BigDecimal;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
public class CreateTransaction {
    private final CurrencyGateway currencyGateway;
    private final TransactionRepository transactionRepository;

    public Output execute(Input input) {
        var currencies = currencyGateway.getCurrencies();
        BigDecimal currency = currencies.get(input.currency);
        var totalAmount = currency.multiply(input.amount);
        var transaction = Transaction.create(input.description, input.amount, input.currency, totalAmount);
        transactionRepository.create(transaction);
        return new Output(transaction.id());
    }

    public record Input(String description, BigDecimal amount, Currency currency) {
    }

    public record Output(String id) {
    }
}

File: kotlin-tomcat/src/main/kotlin/com/ederfmatos/kotlintomcat/usecase/CreateTransaction.kt
package com.ederfmatos.kotlintomcat.usecase

import com.ederfmatos.kotlintomcat.entity.Transaction
import com.ederfmatos.kotlintomcat.enums.Currency
import com.ederfmatos.kotlintomcat.gateway.CurrencyGateway
import com.ederfmatos.kotlintomcat.repository.TransactionRepository
import org.springframework.stereotype.Component
import java.math.BigDecimal

@Component
class CreateTransaction(
    private val currencyGateway: CurrencyGateway,
    private val transactionRepository: TransactionRepository
) {

    fun execute(input: Input): Output {
        val currencies = currencyGateway.getCurrencies()
        val amount = currencies[input.currency]!!.multiply(input.amount)
        val transaction = Transaction.create(
            description = input.description,
            totalAmount = amount,
            amount = input.amount,
            currency = input.currency,
        )
        transactionRepository.create(transaction)
        return Output(transaction.id)
    }

    data class Input(val description: String, val amount: BigDecimal, val currency: Currency)
    data class Output(val id: String)
}